### PR TITLE
BLD: add static to std_c_flags program. Move macosx_arm64 wheel build to cirrus. Add regular macosx_arm64 job (closes #23538, #21765)

### LINKED ISSUE #23538: Nightly wheel build for macosx on arm64 failing
See https://github.com/numpy/numpy/actions/runs/4624671118/jobs/8188472241. The relevant bit seems to be
```
        raise RuntimeError(
    RuntimeError: Broken toolchain during testing C compiler.
    A compiler with support for C99 language features is required.
```
This is new as of 4/5/2023., I'm guessing it is related to the C++ work. @seiko2plus Any ideas.

### PR DESCRIPTION
**Pull request by andyfaff**
#23538 reports that the `macosx_arm64` wheel build is failing. This is because the test program outlined in `std_c_flags` in `setup.py` is not compilable with clang.

The reason for this is outlined in https://github.com/llvm/llvm-project/issues/55274.

This PR makes the test program compilable with clang.

### CODE REVIEW

**Comment by charris** [comment]
close/reopen

**Comment by charris** [comment]
I think actions has been changed and no longer processed serially, so I restarted the tests so to `build` label would have effect.

**Comment by charris** [comment]
Progress. I suspect we will need to check the compiler before setting the flags.

**Comment by andyfaff** [comment]
> Progress. I suspect we will need to check the compiler before setting the flags.

See https://github.com/numpy/numpy/issues/23538#issuecomment-1500748789.

I know how to get the build to complete, remove the `ftrapping-math` flag - it's currently unsupported by the compiler on the `macosx_arm64` target. However, there may need to be further consideration by the numpy team for floating point exception handling on the target.

**Comment by andyfaff** [comment]
Whilst [8a32e35](https://github.com/numpy/numpy/pull/23559/commits/8a32e35327acf2c2ca377658b6528a474565b2f3) builds on my local machine I'm guessing that doesn't work during the CI wheel building matrix entry because that's doing a cross compile. Thus `platform.machine()` (in setup.py) is still x86_64, so `ftrapping-math` still gets added, but the cross-compiler doesn't like the flag. It might be better to build the macosx_arm64 wheels natively on cirrus.

**Comment by andyfaff** [comment]
[fe0357c](https://github.com/numpy/numpy/pull/23559/commits/fe0357c4e58d01c21ff877c6ca9864592c753824)
has had #23550 added.

i.e.
- there is now a native `macosx_arm64` test running on cirrus-ci (with meson)
- the `macosx_arm64` wheels are now natively build (and tested) on cirrus-ci instead of cross-compiled on GHA.

This allows setup.py to be amended to remove `ftrapping-math` on `macosx_arm64`, which means the setup.py based build occurs without issue.

Note that there are still test fails, but these have been highlighted by this PR. The test fail is similar to, if not the same as, #21765.

**Comment by charris** [comment]
Looks like the cp39 macosx arm64 wheel was not built.

The f2py test error should be fixable, see `numpy/f2py/crackfortran.py` Line 2388. I wonder what `platform.machine()` returns on the test machine?

**Comment by andyfaff** [comment]
One of the matrix entries builds two wheels. If the first wheel doesn't finish it's test successfully the second doesn't start. Both will have `platform.machine()==arm64`

**Comment by charris** [comment]
> Both will have platform.machine()==arm64
Here is what the current logic looks like:
```
    machine = platform.machine().lower()
    if machine.startswith(('aarch64', 'power', 'ppc', 'riscv', 's390x', 'sparc')):
        if p <= 20:
            return 16
    else:
        if p < 19:
            return 10
        elif p <= 20:
            return 16
    return -1
```
So you should be able to add a line for arm64. Since Apple M1 doesn't support extended/quad precision, the correct answer should probably be 8 rather than 16 for `aarch64` (quad precision). The test depends on having a Fortran compiler, so that makes me wonder where that compiler comes from and if the real(16) is maybe emulation.

The simplest thing is probably to just add 'arm64' to the list along with 'aarch64', I'm just not convinced that is correct and not a compiler bug :) I assume this is running on real M1 hardware.

**Comment by charris** [comment]
If the test is not running on a real M1, maybe just skip the test with a comment for the apple case.

**Comment by andyfaff** [comment]
I tried compiling and running the following (locally on my M1 Mac):

```fortran
program real_kinds
  integer :: p = selected_real_kind(15)
  print *, p
end program real_kinds
```

for different argument values supplied to `selected_real_kind`. I got the following:

| selected_real_kind | p |
|--------|--------|
| 1 | 4 |
| 2 | 4 |
| 3 | 4 |
| 4 | 4 |
| 5 | 4 | 
| 6 | 4 | 
| 7 | 8 | 
| 15 | 8 |
| 16 | 16 |
| 19 | 16 |
| 20 | 16 |

So it seems to get the test to pass requires changing [L2396](https://github.com/numpy/numpy/blob/main/numpy/f2py/crackfortran.py#L2396) to 

```    
if machine.startswith(('aarch64', 'power', 'ppc', 'riscv', 's390x', 'sparc', 'arm64')):
```

Here are my local machine specs:
```
(dev3) 192-168-1-113:numpy-dev andrew$ gfortran -dumpversion
12.2.0
(dev3) 192-168-1-113:numpy-dev andrew$ gfortran -dumpmachine
aarch64-apple-darwin22
(dev3) 192-168-1-113:numpy-dev andrew$ python
Python 3.9.16 | packaged by conda-forge | (main, Feb  1 2023, 21:38:11) 
Type "help", "copyright", "credits" or "license" for more information.
[Clang 14.0.6 ] on darwin
>>> import platform, sys
>>> sys.platform, platform.machine()
('darwin', 'arm64')
```

The gfortran compiler on my machine was sourced from https://github.com/fxcoudert/gfortran-for-macOS/releases. The gfortran compilers in `gfortran_utils` are repackaged from conda forge, https://github.com/isuruf/gcc/releases/download/gcc-11.3.0-2.

The following program:
```c
#include <stdio.h>
#include <float.h>

int main()
{
    float num1 = FLT_MAX;
    double num2 = DBL_MAX;
    long double num3 = LDBL_MAX;

    printf("%.2f\n", num1);
    printf("%e\n", num2);
    printf("%Le\n", num3);

    return 0;
}
```
gives:
```
340282346638528859811704183484516925440.00
1.797693e+308
1.797693e+308
```

**Comment by charris** [comment]
It is possible that the hardware does support quad precision, but the Apple C compiler doesn't :) In any case, I think the easiest thing to do is add `arm64` to the list. If there are problems, we should hear about them eventually. Of course, NumPy relies on C and won't be able to deal with quad precision, but at least the tests will pass :)

**Comment by andyfaff** [comment]
Also, whilst the default `clang` on M1 may only have 64 bit precision, the gfortran I was referring to above ships with `libquadmath`. AFAIU that's a software emulated quad precision library:

**Comment by andyfaff** [comment]
I tried using the C programs at https://people.math.sc.edu/Burkardt/c_src/gcc_quadmath/gcc_quadmath.html to examine the characteristics of the M1 processor using `docker run -it -v $PWD:/io --platform=linux/arm64 ubuntu`.
When I use the default gcc installed via apt I get:

```
GCC_QUADMATH:
  C version.
  Test the GCC quadmath library.

TEST01
  Using FLOAT arithmetic:
  Compute smallest 1/2^DIV that can be added to 1.
  Number of divisions DIV = 23
  1/2^DIV =         1.192093e-07

TEST02
  Using DOUBLE arithmetic:
  Compute smallest 1/2^DIV that can be added to 1.
  Number of divisions DIV = 52
  1/2^DIV =         2.220446e-16

TEST03
  Using long double arithmetic:
  Compute smallest 1/2^DIV that can be added to 1.
  Number of divisions DIV = 112
  1/2^DIV =         1.925930e-34
```

which leads me to believe that the M1 chip can deal with quad precision, provided the compiler toolchain does. Apple clang does not appear to do so.

**Comment by charris** [comment]
The failing test is unrelated, so this is passing now. The remaining question is the FPU exception trapping, which also seems an Apple Clang thing. I am not inclined to add a lot of Apple specific code to deal with that, so I think the options are:

1. Go with this and hope Apple adds the option at some point 
2. Use another compiler

I'm tending towards 1.  Just to be sure, this also produces wheels for x86_64?

@rgommers, @seberg, @mattip, @seiko2plus Thoughts?

**Comment by charris** [comment]
Let me add @Developer-Ecosystem-Engineering to the notifications.

**Comment by andyfaff** [comment]
> I'm tending towards 1. Just to be sure, this also produces wheels for x86_64?

The original GHA wheels job is still responsible for `macosx_x86_64'

File: numpy/distutils/ccompiler.py [hunk 1]
```diff
@@ -394,8 +395,14 @@ def CCompiler_customize_cmd(self, cmd, ignore=()):
...
                                         cmd.__class__.__name__))
 
-    if hasattr(self, 'compiler') and 'clang' in self.compiler[0]:
+    if (
+        hasattr(self, 'compiler') and
+        'clang' in self.compiler[0] and
+        not (platform.machine() == 'arm64' and sys.platform == 'darwin')
+    ):
         # clang defaults to a non-strict floating error point model.
+        # However, '-ftrapping-math' is not currently supported (2022-04-08)
```
Comment (rgommers) [inline numpy/distutils/ccompiler.py:404]: typo: `2023` rather than `2022`

**Comment by rgommers** [comment]
> 1. Go with this and hope Apple adds the option at some point
> 2. Use another compiler

Definitely (1). Switching away from the default compiler isn't a good idea, and also it doesn't fix anything but rather leaves the problem for users to run into.

**Comment by rgommers** [comment]
This PR looks like it's doing all the right things, +1 for moving the arm64 wheel build to Cirrus.

**Comment by rgommers** [comment]
> Let me add @Developer-Ecosystem-Engineering to the notifications.

The more important thing to weigh in on here is the `-ftrapping-math` support for Clang at https://github.com/numpy/numpy/issues/23538#issuecomment-1500748789.

**Comment by charris** [comment]
Fixed the typo.

**Comment by charris** [comment]
Merging this. Additional comments are welcome.

**Comment by charris** [comment]
Thanks @andyfaff .

**Pull request merged**